Frontend skeleton

dominick-designs opened the pull request:
React web app skeleton tasks: add header, footer, and body. 

* body contains ```react-instant-search``` and is connected with our current test index in algolia. 
* installed tailwindcss
* tailwindcss includes a few config files
        * ```jsconfig.json``` allows for clean path names
        * ```babel-plugin-macros.config``` configures twin.macro
see here for [configuration details for tailwindcss modules](https://owaiskhan.me/post/using-treact-components-custom-react-app)

* ```react-router-dom``` is installed and in place

dominick-designs commented:
I don't understand why it says "Some checks haven't completed yet." Is this a step for devops?

dominick-designs commented:
made fresh commit today which includes: 

- installed DEV dependencies postcss, postcss-cli, autoprefixer, postcss-purgecss
- set up ```postcss.config.js file```
- add script in ```package.json``` file to ```buildtailwind```
- ran ```npm run build``` and saw output in ```frontend``` root
- ran ```npm run start``` and everything appears to work as expected

## NOTE:
- ```@fullhuman/postcss-purgecss``` purges all CSS that is not used; the ```tailwind.css``` file in ```build>build>tailwind.css``` is only @5Kib after using ```purgecss```
- settings in ```postcss.config.js``` should work on firebase; however, see documentation on DEV dependencies if any issues arise. (Additional configuration in ```postcss.config.js``` may be needed?)

anantakrishna commented:
You need to base your branch on top of the latest `main`. Currently your branch is stale, and that is why the CI checks have not been run — they are defined in the commit that is not on your branch.
![image](https://user-images.githubusercontent.com/6065071/97181893-63343600-17ce-11eb-9171-1da20fe54b12.png)

Your branch has conflicts with the main branch, particularly in the README file. I suggest you to revert your changes in that file because it almost duplicate what is already added to it in the `main` branch. You can add frontend-specific details into its own README file inside the `frontend` directory.

There is no need to include “Trello card task” into pull request title and commit messages. Please read the README file in the repository after you rebase your branch on top of `main`.

anantakrishna commented on frontend/.gitignore:
```diff
@@ -21,3 +21,10 @@
 yarn-debug.log*
 yarn-error.log*
+
+# my additions
```
I don't think it's a good wording, because later one will have to dig into the git history to understand whos additions this is. This comment is redundant.

dominick-designs replied:
OK, I will remove this heading.

anantakrishna commented on frontend/.gitignore:
```diff
@@ -21,3 +21,10 @@
 yarn-error.log*
+
+# my additions
+notes.md
```
What is the rationale behind ignoring this file? I propose to at least use some hidden file or something like `notes.local.md`.

dominick-designs replied:
OK, I will name it as suggested. This is a local file for my notes.

anantakrishna commented on frontend/.gitignore:
```diff
@@ -21,3 +21,10 @@
+
+# my additions
+notes.md
+../.DS_Store
+./.DS_Store
```
I guess this can be added to the root `gitignore`.

anantakrishna commented on frontend/package.json:
```diff
@@ -6,13 +6,29 @@
   },
   "scripts": {
-    "start": "react-scripts start",
+    "start": "BROWSER='firefox developer edition' react-scripts start",
```
Are you sure it should be a part of the source code? What if other developer don't have this browser (like me)? Maybe there is another way to specify it locally?

dominick-designs replied:
I am not sure how to do this. I don't have any experience rebasing or reverting. Can you give me a link on how what to do? 

> You need to base your branch on top of the latest `main`. Currently your branch is stale, and that is why the CI checks have not been run — they are defined in the commit that is not on your branch.
> ![image](https://user-images.githubusercontent.com/6065071/97181893-63343600-17ce-11eb-9171-1da20fe54b12.png)
> 
> Your branch has conflicts with the main branch, particularly in the README file. I suggest you to revert your changes in that file because it almost duplicate what is already added to it in the `main` branch. You can add frontend-specific details into its own README file inside the `frontend` directory.
> 
> There is no need to include “Trello card task” into pull request title and commit messages. Please read the README file in the repository after you rebase your branch on top of `main`.

anantakrishna commented on frontend/package.json:
```diff
@@ -6,13 +6,29 @@
-    "start": "react-scripts start",
+    "start": "BROWSER='firefox developer edition' react-scripts start",
     "build": "react-scripts build",
+    "buildtailwind": "postcss src/styles/tailwind.css -o public/build/tailwind.css",
```
If this script has to be called before/after main build process, I'd prefer you to include it into the build script itself — for the sake of automated build. Otherwise we'll have to adjust the DevOps workflow, which is possible but not desirable.

dominick-designs replied:
OK this script is removed.

anantakrishna commented on frontend/public/index.html:
```diff
@@ -25,6 +25,8 @@
     -->
     <title>React App</title>
+
+    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/instantsearch.css@7/themes/algolia-min.css">
```
In https://www.algolia.com/doc/guides/building-search-ui/widgets/customize-an-existing-widget/react/#loading-the-theme section there are several methods described. Why have you chosen CDN way instead of npm & Webpack? Just curious.

dominick-designs replied:
```import 'instantsearch.css/themes/algolia.css';``` is now in ```App.js``` and works as expected. Removed ```CDN``` link in ```index.html```

anantakrishna commented on frontend/src/components/algolia/algoliaInstantSearch.js:
```diff
@@ -0,0 +1,87 @@
+import React, { Component } from 'react';
+import algoliasearch from 'algoliasearch/lite';
+import {
+  Highlight, Hits, InstantSearch,
+
+
+  Pagination, SearchBox
+} from 'react-instantsearch-dom';
```
Why vertical space inside import? There is prettier configured for the whole project in the root folder. If you use VSCode it should format on save…

anantakrishna commented on frontend/src/components/algolia/algoliaInstantSearch.js:
```diff
@@ -0,0 +1,87 @@
+
+
+
+const searchClient = algoliasearch(
+  'A0IP7ZFQR8',
+  '7d853e1791e8f7668cb942eeb76888ac'
+);
```
Of course we should not commit API keys into repo, even test ones. Move this into the local (ignored) env file and use variables here.

dominick-designs replied:
created .env.local file and this works locally. 

## NOTE from ```create-react-app``` docs: 

> WARNING: Do not store any secrets (such as private API keys) in your React app!

> Environment variables are embedded into the build, meaning anyone can view them by inspecting your app's files.

https://create-react-app.dev/docs/adding-custom-environment-variables/

After reading this, I am not sure if I should push this file? Since it is ignored it will not be pushed. But the docs say that after build these values will be exposed. This is why Firebase probably has a place to enter thise values.

anantakrishna replied:
You need not pushing this file. The proper values of these variables will be provided to React in our deployment pipeline, so they will be embedded, yes. That's not a problem. But we should not commit them into the repository.

dominick-designs replied:
If I don't push the file in this section will not be visible. I have inserted the environment variable and it works as expected. 

do you mean I should not push the local environment variable file?￼

anantakrishna replied:
Yes I mean not to push the local env file. However, please describe needed variables in the frontend's README file. Similar to this: https://github.com/PureBhaktiArchive/audioseva#frontend-environment-variables

anantakrishna commented on frontend/src/helpers/EmbedIframe.js:
```diff
@@ -0,0 +1,21 @@
+  // this is google contact form form iframe
+  const FlexDiv = tw.div`flex flex-col items-center lg:block h-full `;
+
+  url = "https://docs.google.com/forms/d/e/1FAIpQLScIUS0rJD1UkPshb-72MoFV9UN_m7CnrCAcYwgumFEiXCM_Sg/viewform?embedded=true"
```
Why this? I don't expect us to embed google form at all.

dominick-designs replied:
Removed. This was somehow place from a previous project. I will clean up the code as I see there are references to other files that don't exist.

anantakrishna commented on frontend/src/styles/App.css:
```diff
@@ -0,0 +1,60 @@
+.header-title {
+  font-size: 1.2rem;
+  font-weight: normal;
+}
+
+.header-title::after {
+  content: ' ▸ ';
+  padding: 0 0.5rem;
+}
+
+.header-subtitle {
+  font-size: 1.2rem;
+}
+
+.container {
+  max-width: 1200px;
+  margin: 0 auto;
+  padding: 1rem;
+}
+
+.search-panel {
+  display: flex;
+}
+
+.search-panel__filters {
+  flex: 1;
+}
+
+.search-panel__results {
+  flex: 3;
+}
+
+.searchbox {
+  margin-bottom: 2rem;
+}
+
+.pagination {
+  margin: 2rem auto;
+  text-align: center;
+}
```
How this `App.css` lives together with Tailwind CSS?

dominick-designs replied:
this was part of the Algolia demo project. This can be replaced with styled-components using tailwindcss. Note, tailwindcss allows interaction with other CSS files; this works as expected but will be refactored.

anantakrishna replied:
I see. Yes, it's better to refactor to use Tailwind CSS.

anantakrishna reviewed (changes requested):
Here is the per-file review.

dominick-designs commented:
Failed build because of environment variables. Let me know how to fix this. 

Thank you,

anantakrishna commented:
I will configure variables in the CI script, you don't need to do anything about it. I will review the changes today later.

anantakrishna commented on .vscode/settings.json:
```diff
@@ -8,7 +8,7 @@
   },
   "editor.codeActionsOnSave": {
     "source.fixAll.eslint": true,
-    "source.organizeImports": true
```
Can you comment for what reason you have removed this setting? I find it useful to automatically organize imports on save. Sometimes it removes unused imports which you are going to use, and that causes some inconvenience. However if we remove this, then we will just forget to trigger this source action manually, which will lead to chaos in imports.

dominick-designs replied:
I was noticing that my imports were being automatically ordered alphabetically and I thought that ```import React from "react"``` had to be at the top. I was also troubleshooting an issue and changed this value to ```false``` as part of troubleshooting steps; I didn't change it back to ```true``` after resolving my troubleshooting. 

I didn't realize that my local vscode settings would be included in a commit. I am willing to accept any pull that includes modifications to vscode settings.

anantakrishna replied:
The point is that you have modified the workspace settings, which are stored in the repository in `.vscode/settings.json`. These settings are supposed to be committed, but the changes should be reasonable. That is why I request to use any graphical GIT tool like the one included into VSCode, because it allows to inspect every change that you are going to commit. Please read the “Don't include irrelevant changes into commits.” section of the README file.

anantakrishna commented on .vscode/settings.json:
```diff
@@ -43,5 +43,14 @@
   },
-  "formatFiles.excludePattern": "**/*.min.js"
+  "formatFiles.excludePattern": "**/*.min.js",
+  "prettier.eslintIntegration": true,
```
This setting seems outdated and deprecated: https://github.com/prettier/prettier-vscode/issues/999

And it is not recognized in my VSCode even:
![image](https://user-images.githubusercontent.com/6065071/97592610-6c720c80-1a33-11eb-8c7a-6feda3298d07.png)

dominick-designs replied:
I am not sure about this either. I don't even know if I modified this. I didn't realize that my local vscode settings would be included in a commit. I am not sure how to fix this. I am willing to accept any pull that includes modifications to vscode settings.

anantakrishna replied:
Just revert all unnecessary changes and make another commit.

anantakrishna replied:
In regard to `prettier.eslintIntegration`, I share the official prettier's point of view:

- https://prettier.io/docs/en/comparison.html - “use Prettier for formatting and linters for catching bugs!”
- https://prettier.io/docs/en/integrating-with-linters.html, the first section only.
- https://prettier.io/docs/en/install.html - Install `prettier` locally into the frontend. Use `eslint-config-prettier` to make Prettier and ESLint play nice together.

anantakrishna commented on .vscode/settings.json:
```diff
@@ -43,5 +43,14 @@
+
+    "enable": true
+  },
+  "eslint.nodeEnv": "",
+  "eslint.options": {
+  
+  }
```
What is the use of this settings block?

dominick-designs replied:
I am not sure. I didn't realize that my local vscode settings would be included in a commit. I am not sure how to fix this. I am willing to accept any pull that includes modifications to vscode settings.

anantakrishna commented on frontend/.gitignore:
```diff
@@ -13,6 +13,7 @@
 
 # misc
 .DS_Store
+.env*.local
```
Isn't this redundant due to the explicit listing below?

dominick-designs replied:
Yes, this can be seen as redundant. I just wanted to make sure so. This can be removed.

anantakrishna commented on frontend/.gitignore:
```diff
@@ -21,3 +22,8 @@
+
+notes.local.md
+/build
+/public/build
```
Do we have `public/build` at all?

dominick-designs replied:
regarding ```public/build``` this can be removed. This is inserted during testing with tailwindcss

anantakrishna reviewed (changes requested):
Besides in-file remarks, please accept these points:

- I have added environment variables to the CI script, now tests start successfully, but the test itself fails because you have changed the main page. Let's change the test so that it succeeds. Don't make it too complicated because the main page will be changed a lot soon.
- There are many inconsistencies in the files, like extra whitespaces etc. Make sure that each time you save a file, prettier in VSCode automatically formats the files. It should work provided you open the root folder and have installed the recommended VSCode extensions.

Note: I have added commits to the same branch, so pull before adding your commits.

dominick-designs commented:
> * There are many inconsistencies in the files, like extra whitespaces etc. Make sure that each time you save a file, prettier in VSCode automatically formats the files. It should work provided you open the root folder and have installed the recommended VSCode extensions.

I think part of the issue is that I was confused about prettier. I had been working with ```prettier+``` and didn't realize it was different until a fews days after working on this branch. I have uninstalled ```prettier +``` and installed ```prettier``` I am sure if I need to make modifications to the prettier settings in order to remove inconsistencies in the files like you mentioned above.

dominick-designs commented:
> Besides in-file remarks, please accept these points:
> 
> * I have added environment variables to the CI script, now tests start successfully, but the test itself fails because you have changed the main page. Let's change the test so that it succeeds. Don't make it too complicated because the main page will be changed a lot soon.
> * There are many inconsistencies in the files, like extra whitespaces etc. Make sure that each time you save a file, prettier in VSCode automatically formats the files. It should work provided you open the root folder and have installed the recommended VSCode extensions.
> 
> Note: I have added commits to the same branch, so pull before adding your commits.

I have made a successful ```pull``` and addressed these issues. Should I commit now? I want to make sure I make meaningful commits rather than committing after making only a few modifications on the settings and configurations.

anantakrishna commented:
> I think part of the issue is that I was confused about prettier. I had been working with `prettier+` and didn't realize it was different until a fews days after working on this branch. I have uninstalled `prettier +` and installed `prettier`

I'm sure prettier+ is using the same config file as the `esbenp.prettier-vscode` does. Hence, it should have been working for you…

By the way, there are two Prettier+ extensions: `svipas.prettier-plus` and `shunqian.prettier-plus`. Which one are you using? And do you feel any advantage of it over `esbenp.prettier-vscode`? Maybe we should both use Prettier+. Let's discuss.

> I am sure if I need to make modifications to the prettier settings in order to remove inconsistencies in the files like you mentioned above.

I believe the current prettier options are reasonable. However, I'm open to any suggestion, feel free to propose any change with rationale: https://prettier.io/docs/en/options.html.

> Should I commit now? I want to make sure I make meaningful commits rather than committing after making only a few modifications on the settings and configurations.

Usually it's all right to make small commits addressing particular issues. Anyhow the pull request will be squashed into one commit to `main`, so small commits are just for convenience while working on the pull request.

dominick-designs commented:
I was using this ```prettier+``` extension
> svipas.prettier-plus

To be honest I installed this thinking it was the original ```prettier``` and only realized that the ```+``` distinguished it from the original while working on this project. I am not sure about the differences. The developer states, 

> "Prettier (code formatter) for the VS Code. To see the difference between official Prettier extension you can take a look at the CHANGELOG." 

I just made another commit but correct me if the vscode settings are incorrect. I've never worked with this directory before. I don't recall seeing this ```.vscode``` directory in any of my other projects so I am not familiar with these ```key > value``` pairs.

anantakrishna commented:
Let me adjust the VSCode settings and prettier configuration myself, I will add commits here. I guess after we set the skeleton up, further development will go smoother.

dominick-designs commented:
ok sounds good. I know it will go smoother since this is really my first time working on this workflow. Thank you for your patience. 



> On Oct 30, 2020, at 17:54, Ananta Krsna dasa <notifications@github.com> wrote:
> 
> 
> Let me adjust the VSCode settings and prettier configuration myself, I will add commits here. I guess after we set the skeleton up, further development will go smoother.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/PureBhaktiArchive/archive-web-app/pull/3#issuecomment-719878867>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAJWRDJ2FXTONOBABR7YDILSNOC6PANCNFSM4S53ZFQA>.
>

anantakrishna commented:
It's done. I have also formatted all the files with prettier automatically. Now please pull the changes and fix the test, as it still fails. After that we can accept the pull request.

dominick-designs commented:
Is there a way to run the tests locally? I am not sure what tests are failing or how to address the failures. 



> On Oct 30, 2020, at 19:37, Ananta Krsna dasa <notifications@github.com> wrote:
> 
> 
> It's done. I have also formatted all the files with prettier automatically. Now please pull the changes and fix the test, as it still fails. After that we can accept the pull request.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/PureBhaktiArchive/archive-web-app/pull/3#issuecomment-719888173>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAJWRDJ7FSBXJGMN6T5AWNDSNOPCPANCNFSM4S53ZFQA>.
>

anantakrishna commented:
Sure - https://create-react-app.dev/docs/running-tests/#command-line-interface.

There is the only one test added by the `create-react-app` automatically from the template. Please get familiar with the testing approach used in Create React App and just fix the test in an easy way as of now. I'd prefer us to add some tests later to stay safe while introducing changes. As of now, just fix this one test.
